src/tools/SwitchingFunction.cpp

```diff
@@ -304,259 +646,232 @@ void SwitchingFunction::set(const std::string & definition,std::string& errormsg
-  if(type==rational) {
-    ostr<<" nn="<<nn<<" mm="<<mm;
-  } else if(type==nativeq) {
-    ostr<<" beta="<<beta<<" lambda="<<lambda<<" ref="<<ref;
-  } else if(type==smap) {
-    ostr<<" a="<<a<<" b="<<b;
-  } else if(type==cubic) {
-    ostr<<" dmax="<<dmax;
-  } else if(type==leptontype) {
-    ostr<<" func="<<lepton_func;
+void SwitchingFunction::set(const std::string & definition,std::string& errormsg) {
+  std::vector<std::string> data=Tools::getWords(definition);
+#define CHECKandPARSE(datastring,keyword,variable,errormsg) \
+  if(Tools::findKeyword(datastring,keyword) && !Tools::parse(datastring,keyword,variable))\
+    errormsg="could not parse " keyword; //adiacent strings are automagically concatenated
+#define REQUIREDPARSE(datastring,keyword,variable,errormsg) \
+  if(!Tools::parse(datastring,keyword,variable))\
+    errormsg=keyword " is required for " + name ; //adiacent strings are automagically concatenated
 
+  if( data.size()<1 ) {
+    errormsg="missing all input for switching function";
+    return;
   }
-  return ostr.str();
-}
+  std::string name=data[0];
+  data.erase(data.begin());
+  double r0=0.0;
+  double d0=0.0;
+  double dmax=std::numeric_limits<double>::max();
+  init=true;
+  CHECKandPARSE(data,"D_0",d0,errormsg);
+  CHECKandPARSE(data,"D_MAX",dmax,errormsg);
 
-double SwitchingFunction::do_rational(double rdist,double&dfunc,int nn,int mm)const {
-  double result;
-  if(2*nn==mm) {
-// if 2*N==M, then (1.0-rdist^N)/(1.0-rdist^M) = 1.0/(1.0+rdist^N)
-    double rNdist=Tools::fastpow(rdist,nn-1);
-    double iden=1.0/(1+rNdist*rdist);
-    dfunc = -nn*rNdist*iden*iden;
-    result = iden;
+  bool dostretch=false;
+  Tools::parseFlag(data,"STRETCH",dostretch); // this is ignored now
+  dostretch=true;
+  bool dontstretch=false;
+  Tools::parseFlag(data,"NOSTRETCH",dontstretch); // this is ignored now
+  if(dontstretch)
+    dostretch=false;
+  if(name=="CUBIC") {
+    //cubic is the only switch type that only uses d0 and dmax
+    function = PLMD::Tools::make_unique<switchContainers::cubicSwitch>(d0,dmax);
   } else {
-    if(rdist>(1.-100.0*epsilon) && rdist<(1+100.0*epsilon)) {
-      result=nn/mm;
-      dfunc=0.5*nn*(nn-mm)/mm;
+    REQUIREDPARSE(data,"R_0",r0,errormsg);
+    if(name=="RATIONAL") {
+      int nn=6;
+      int mm=0;
+      CHECKandPARSE(data,"NN",nn,errormsg);
+      CHECKandPARSE(data,"MM",mm,errormsg);
+      function = switchContainers::rationalFactory(d0,dmax,r0,nn,mm);
+      //function = PLMD::Tools::make_unique<switchContainers::rationalSwitch>(d0,dmax,r0,nn,mm);
+    } else if(name=="SMAP") {
+      int a=0;
+      int b=0;
+      //in the original a and b are "default=0",
+      //but you divide by a and b during the initialization!
+      //better an error message than an UB, so no default
+      REQUIREDPARSE(data,"A",a,errormsg);
+      REQUIREDPARSE(data,"B",b,errormsg);
+      function = PLMD::Tools::make_unique<switchContainers::smapSwitch>(d0,dmax,r0,a,b);
```

Comment:
@GiovanniBussi here I changed the behavior because I saw division by `a` and `b`
Do you think that this could broke something in some egg?

Reply:
I don't think so, I guess results would make no sense if one used either a or b = 0